test/Libraries/Microsoft.Extensions.Caching.Hybrid.Tests/TestEventListener.cs

```diff
@@ -0,0 +1,178 @@
+                                        break;
+                                    case "Mean":
+                                        isIncrement = false;
+                                        value = Convert.ToDouble(pair.Value);
+                                        break;
+                                    case "Increment":
+                                        isIncrement = true;
+                                        value = Convert.ToDouble(pair.Value);
+                                        break;
+                                }
+                            }
+
+                            if (name is not null && value is not null)
+                            {
+                                if (isIncrement && _counters.TryGetValue(name, out var oldPair))
+                                {
+                                    value += oldPair.value; // treat as delta from old
+                                }
+
+                                Debug.WriteLine($"{name}={value}");
+                                _counters[name] = (displayName, value.Value);
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    _events.Add((eventData.EventId, eventData.EventName ?? "", eventData.Level));
+                }
+            }
+        }
+
+        base.OnEventWritten(eventData);
+    }
+
+    public (int id, string name, EventLevel level) SingleEvent()
+    {
+        (int id, string name, EventLevel level) evt;
+        lock (SyncLock)
+        {
+            evt = Assert.Single(_events);
+        }
+
+        return evt;
+    }
+
+    public void AssertSingleEvent(int id, string name, EventLevel level)
+    {
+        var evt = SingleEvent();
+        Assert.Equal(name, evt.name);
+        Assert.Equal(id, evt.id);
+        Assert.Equal(level, evt.level);
+    }
+
+    public double AssertCounter(string name, string displayName)
+    {
+        lock (SyncLock)
+        {
+            Assert.True(_counters.TryGetValue(name, out var pair), $"counter not found: {name}");
+            Assert.Equal(displayName, pair.displayName);
+
+            _counters.Remove(name); // count as validated
+            return pair.value;
+        }
+    }
+
+    public void AssertCounter(string name, string displayName, double expected)
+    {
+        var actual = AssertCounter(name, displayName);
+        if (!Equals(expected, actual))
+        {
+            Assert.Fail($"{name}: expected {expected}, actual {actual}");
+        }
+    }
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Major Bug", "S1244:Floating point numbers should not be tested for equality", Justification = "Test expects exact zero")]
+    public void AssertRemainingCountersZero()
+    {
+        lock (SyncLock)
+        {
+            foreach (var pair in _counters)
+            {
+                if (pair.Value.value != 0)
```

Comment:
nit: `!double.IsZero(pair.Value.value)` probably gets rid of the warning?

Reply:
`double.IsZero` does not appear to be a thing